tools/foreman-cockpit-session

```diff
@@ -0,0 +1,254 @@
+#!/usr/bin/env ruby
+
+require "json"
+require "net/https"
+
+# Cockpit protocol, encoding and decoding of control messages.
+
+def send_control(msg)
+  text = JSON.dump(msg)
+  $stdout.write("#{text.length+1}\n\n#{text}")
+  $stdout.flush
+end
+
+def read_control
+  size = $stdin.readline.chomp.to_i
+  raise ArgumentError, "Invalid frame: invalid size" if size.zero?
+  data = $stdin.read(size)
+  raise ArgumentError, "Invalid frame: too short" if data == nil || data.length < size
+  JSON.parse(data)
+end
+
+# Specific control messages
+
+def send_auth_challenge(challenge)
+  send_control({ "command" => "authorize",
+                 "cookie" => "1234", # must be present, but value doesn't matter
+                 "challenge" => challenge
+               })
+end
+
+def send_auth_response(response)
+  send_control({ "command" => "authorize",
+                 "response" => response
+               })
+end
+
+def read_auth_reply
+  cmd = read_control()
+  response = cmd["response"]
+  raise ArgumentError, "Did not receive a valid authorize command" if cmd["command"] != "authorize" || !response
+  response
+end
+
+def exit_with_problem(problem, message, auth_methods)
+  send_control({ "command" => "init",
+                 "problem" => problem,
+                 "message" => message,
+                 "auth-method-results" => auth_methods
+               })
+  exit 1
+end
+
+# Talking to Foreman
+
+def get_token_from_auth_data(auth_data)
+  auth_data.split(" ")[1]
+end
+
+def foreman_call(path, token)
+  foreman = ENV["FOREMAN_URL"] || "https://localhost/"
+  uri = URI(foreman + path)
+  req = Net::HTTP::Get.new(uri)
+  req["Cookie"] = "_session_id=#{token}"
+  res = Net::HTTP.start(uri.hostname, uri.port,
+                        :use_ssl => uri.scheme == "https", :verify_mode => OpenSSL::SSL::VERIFY_NONE) {
+    |http| http.request(req)
+  }
+
+  case res.code
+  when "200"
+    return JSON.parse(res.body)
+  when "401"
+    exit_with_problem("authentication-failed",
+                      "Token was not valid",
+                      { "password" => "not-tried", "token" => "denied" })
+  when "404"
+    return nil
+  else
+    $stderr.write("Error talking to foreman: #{res.body}\n")
```

Comment:
It would be good to wrap the `$stderr` into `log` method, or something like that. Also, I would recommend `$stderr.puts("..."`)`. over `$stderr.write("...\n")`

Reply:
Also, the ideal would be to create `logger = Logger.new($stderr)` and allow setting log level via setting and add debug logging as much as possible (including all requests and responses) once something doesn't work, it's pretty hard to figure out what's going on.